Source/DafnyCore/Resolver/PreType/PreTypeToType.cs

```diff
@@ -0,0 +1,174 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using Microsoft.Boogie;
+
+namespace Microsoft.Dafny;
+
+
+class PreTypeToTypeVisitor : ASTVisitor<IASTVisitorContext> {
+  public override IASTVisitorContext GetContext(IASTVisitorContext astVisitorContext, bool inFunctionPostcondition) {
+    return astVisitorContext;
+  }
+
+  public PreTypeToTypeVisitor() {
+  }
+
+  public static Type PreType2Type(PreType preType) {
+    var pt = (DPreType)preType.Normalize(); // all pre-types should have been filled in and resolved to a non-proxy
+    if (pt.PrintablePreType != null) {
+      pt = pt.PrintablePreType;
+    }
+    switch (pt.Decl.Name) {
+      case "bool":
+        return Type.Bool;
+      case "char":
+        return Type.Char;
+      case "int":
+        return Type.Int;
+      case "real":
+        return Type.Real;
+      case "ORDINAL":
+        return Type.BigOrdinal;
+      case "set":
+        return new SetType(true, PreType2Type(pt.Arguments[0]));
+      case "iset":
+        return new SetType(false, PreType2Type(pt.Arguments[0]));
+      case "multiset":
+        return new MultiSetType(PreType2Type(pt.Arguments[0]));
+      case "seq":
+        return new SeqType(PreType2Type(pt.Arguments[0]));
+      case "map":
+        return new MapType(true, PreType2Type(pt.Arguments[0]), PreType2Type(pt.Arguments[1]));
+      case "imap":
+        return new MapType(false, PreType2Type(pt.Arguments[0]), PreType2Type(pt.Arguments[1]));
+      default:
```

Comment:
no bv? string has already been desugared into seq<char>? nat treated as int?

Reply:
Since pre-types don't include type synonyms (like `string`) or subset types (like `nat`), pre-type inference only comes up with types like `seq<char>` and `int`. Those are handled in this method.

Awkwardly, the `Type` hierarchy has special nodes for several different types. For example, `seq` is represented by a `SeqType` node. (This would be good to change in the future.) The pre-type hierarchy is simpler, treating all non-proxy pre-types as (what the literature would call) type constructors.

Bitvectors are represented as `ValuetypeDecl`s. To create such a type from a pre-type, the `ValuetypeDecl.CreateType` method is called.